File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int,int>> mp;
    int find_target(vector<int> &nums, int sum, int target, int idx){
        if(idx == nums.size()){
            if(sum == target) return 1;
            return 0;
        }
        
        if(mp.count(idx) && mp[idx].count(sum)){
            return mp[idx][sum];
        }
        
        int ans = 0;
        ans += find_target(nums, sum - nums[idx], target, idx + 1);
        ans += find_target(nums, sum + nums[idx], target, idx + 1);
        
        return mp[idx][sum] = ans;
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        return find_target(nums, 0, target, 0);
    }
};

File: 0494-target-sum/0494-target-sum.java
class Solution {
    private Map<Integer, Map<Integer, Integer>> mp;

    private int findTarget(int[] nums, int sum, int target, int idx) {
        int n = nums.length;
        if(idx == n) {
            if(sum == target) return 1;
            else return 0;
        }

        if(mp.containsKey(idx) && mp.get(idx).containsKey(sum)) {
            return mp.get(idx).get(sum);
        }

        int ans = 0;
        mp.putIfAbsent(idx, new HashMap<>());
        ans += findTarget(nums, sum - nums[idx], target, idx + 1);
        ans += findTarget(nums, sum + nums[idx], target, idx + 1);
        
        mp.get(idx).put(sum, ans);
        return ans;
    }

    public int findTargetSumWays(int[] nums, int target) {
        mp = new HashMap<>();
        return findTarget(nums, 0, target, 0);
    }
}